feat: Добавление режиссёров в фильмы.

ChesnokovEK opened the pull request.

4echo committed:
feat: Добавление режиссёров в фильмы.

4echo committed:
feat: Добавление режиссёров в фильмы.

SeRion525 commented on src/main/java/ru/yandex/practicum/filmorate/repository/film/JdbcFilmRepository.java:
```diff
@@ -5,28 +5,33 @@
     private static final String GET_ALL_QUERY = """
-            SELECT films.*, mpa.name AS mpa_name, genres.genre_id, genres.name AS genre_name FROM films
+            SELECT films.*, mpa.name AS mpa_name, genres.genre_id, genres.name AS genre_name,
+            directors.director_id as director_id, directors.name as director_name FROM films
```
столбец directors.director_id необязательно явно переименовывать в directors_id, оно и так будет отображаться нужным именем

ChesnokovEK replied:
Поправил

Kuras08 commented on src/main/java/ru/yandex/practicum/filmorate/controller/DirectorController.java:
```diff
@@ -0,0 +1,52 @@
+    @Validated({Default.class, Update.class})
+    public Director updateDirector(@RequestBody Director director) {
+        directorService.update(director);
+        return directorService.getDirectorById(director.getId());
```
метод update должен сам должен возвращать режиссера, используя getDirectorById,  два раза делам запрос в базу данных, это не очень эффективно.
дожно быть просто
return directorService.update(director);

ChesnokovEK replied:
Поправил

SeRion525 commented on src/main/java/ru/yandex/practicum/filmorate/repository/film/JdbcFilmRepository.java:
```diff
@@ -138,6 +188,19 @@ public List<Film> getMostPopular(int count) {
 
+    @Override
+    public List<Film> filmsByDirector(long directorId, String sortBy) {
+        List<Film> films;
```
Необязательно создавать переменную для списка, лучше сразу возращать его в return при вызове метода findMany

ChesnokovEK replied:
Поправил

SeRion525 commented on src/main/java/ru/yandex/practicum/filmorate/repository/film/JdbcFilmRepository.java:
```diff
@@ -138,6 +188,19 @@ public List<Film> getMostPopular(int count) {
+    @Override
+    public List<Film> filmsByDirector(long directorId, String sortBy) {
+        List<Film> films;
+        if (sortBy.equals("year".toLowerCase())) {
```
Валидацию параметра sortBy лучше проводить в контроллере или сервисе, а получение фильмов режиссёра в репозитории разбить на два метода, по году и по лайкам, нужный метод вызывать в сервисе в зависимости от параметра сортировки

ChesnokovEK replied:
Методы написал, выбор метода перенес в сервис

Kuras08 commented on src/main/java/ru/yandex/practicum/filmorate/model/Director.java:
```diff
@@ -0,0 +1,13 @@
+
+@Getter
+@Setter
+@RequiredArgsConstructor
```
@RequiredArgsConstructor — это аннотация из библиотеки Lombok, которая автоматически генерирует конструктор для класса с полями, помеченными как final, а также с полями, которые имеют модификатор @NonNull. https://www.baeldung.com/java-lombok-constructor-annotation

ChesnokovEK replied:
Поправил

SeRion525 commented on src/main/java/ru/yandex/practicum/filmorate/service/BaseDirectorService.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @Override
+    public Director deleteDirectorById(long id) {
```
Не уверен, что при удалении нужно возвращать удалённую сущность

ChesnokovEK replied:
Скорее всего я хотел её возвращать для логов, но т.к. оно пока нигде не используется, поменял на void

SeRion525 commented on src/main/java/ru/yandex/practicum/filmorate/service/BaseFilmService.java:
```diff
@@ -101,6 +106,12 @@ public List<Film> getMostPopular(int count) {
     }
 
+    @Override
+    public List<Film> filmsByDirector(long directorId, String sortBy) {
```
Сюда перенести выбор сортировки и нужного метода в репозитории

SeRion525 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -69,4 +69,9 @@ public void removeLike(@PathVariable @Positive long id, @PathVariable @Positive
     }
+
+    @GetMapping("/director/{directorId}")
+    public List<Film> filmsByDirector(@PathVariable long directorId, @RequestParam String sortBy) {
```
Лучше здесь проводить валидацию параметра sortBy

SeRion525 commented on src/main/resources/schema.sql:
```diff
@@ -44,4 +44,16 @@ CREATE TABLE IF NOT EXISTS likes
+
+CREATE TABLE IF NOT EXISTS director_films
+(
+    film_id BIGINT NOT NULL REFERENCES films(film_id),
```
Столбцам с ссылками на режиссёра лучше задать правило ON DELETE CASCADE, чтобы при удалении режиссёра, удалялись строки в которых используется данный id.

ChesnokovEK replied:
Как будто бы не только для режиссёра стоит это делать? Взял на себя смелось и дописал ON DELETE CASCADE везде, где оно на мой взгляд имело смысл (по факту - везде, кроме mpa_id у фильма... хотя возможно туда тоже стоит добавить)

SeRion525 commented on src/main/resources/schema.sql:
```diff
@@ -44,4 +44,16 @@ CREATE TABLE IF NOT EXISTS likes
+CREATE TABLE IF NOT EXISTS director_films
+(
+    film_id BIGINT NOT NULL REFERENCES films(film_id),
+    director_id BIGINT NOT NULL REFERENCES directors(director_id)
```
Столбцам с ссылками на режиссёра лучше задать правило ON DELETE SET NULL, чтобы при удалении режиссёра столбец director_id у фильма был NULL

SeRion525 reviewed (commented):
Мама, я ревьюер :3

4echo committed:
fix: Изменения после командного ревью.

- Убран дублирующий запрос в БД в контроллере в методе updateDirector()
- Убрана ненужная аннотация в модели Director
- Валидация параметра запроса (sortBy) вынесена из репозитория в сервис
- Поправлены лишние объявления в коде
- В схему БД добалены каскадные удаления записей
